refactor 'list_permute_term_gen_inner'

ulugbekna opened the pull request:
Changing the implementation of the list_permute_term_gen_inner helped.

In my numerous testings (including long test suites of 5000 tests), both CI tests and compiler tester are quite fast.

make tests args="--long" (5000 tests) runs in ~300s, which is 0.06s per test.
make exec args="--long" (5000 tests) runs in ~2m, which is 0.25s per test.

These measurements vary from seed to seed. (seed 448388817 is very slow: ~154s on make tests)

gasche commented on src/efftester.ml:
```diff
@@ -1089,39 +1089,57 @@ module Generators (Ctx : Context) = struct
+      other rules if this one fails*)
+    let gen_freq lst rand_k =
+      let rand_k = rand_k - 1 in
+      let rec aux lst (acc_lst, acc_sum) =
```
I'm not fond of having two functions named `aux`. What about `pick` for this one, and `pick_all` for the second one?

ulugbekna replied:
Yes, a more declarative naming would be more appropriate.

gasche commented on src/efftester.ml:
```diff
@@ -1089,39 +1089,57 @@ module Generators (Ctx : Context) = struct
+      we need the picked element's rule to try to generate a value and the list to have
+      other rules if this one fails*)
+    let gen_freq lst rand_k =
+      let rand_k = rand_k - 1 in
```
I find this definition a bit confusing. Maybe you could *not* decrement `rand_k` here and use a `<=`-test below?

gasche commented on src/efftester.ml:
```diff
@@ -1089,39 +1089,57 @@ module Generators (Ctx : Context) = struct
+        | [] -> failwith "gen_term_from_rules: gen_freq: too large rand_k"
+        | ((w, _g) as curr) :: rest ->
+          if rand_k < acc_sum + w
+          then (curr, List.rev acc_lst @ rest)
```
This is a minor point, but in this situation you can use `List.rev_append acc_lst rest` which is faster. Note that in fact you wouldn't need to reverse `acc_lst` -- it's fine to reorder the list as long as the total weight remains the same -- but `List.rev_append` is slightly faster than `List.append` and also preserves the order, so this is a win.

gasche commented:
This is a fun problem so I also wrote my own version of `list_permute_gen`, trying to get an algorithm that is as close to linear-time as possible. (The original `list_permute_term_gen implementation` was quadratic. The implementation that you propose is much faster but it remains quadratic in the worst case, due to the fact that picking an element requires removing it from the list, which is a linear operation.) My code is included below, and I believe that it has O(n log n) complexity. That said, it is probably *slower* for small lists, given that it builds intermediate structures.

```ocaml
type 'a weighted = int * 'a
let weight (w, v) = w

(* A weighted tree carries weighted elements. The weight of the tree
   is the total sum of weights of its elements. Internal nodes store
   the weights of their children, so that the weight of a node can always
   be computed in constant time. *)
type 'a weighted_tree = 'a tree weighted

and 'a tree =
  | Nothing
  | Leaf of 'a
  | Node of 'a weighted_tree * 'a weighted_tree

let nothing = (0, Nothing)
let leaf (w, v) = (w, Leaf v)
let node ((w1, _) as t1) ((w2, _) as t2) = (w1 + w2, Node (t1, t2))

(* [extract i tree] returns:
   - the element (w, v) of the tree such that
     the total weight wl ot the elements to its left satisfies (wl <= i < wl + v)
   - a tree with this element replaced by a Nothing node,
     with the weights updated

  If the input tree is well-balanced, this runs in logarithmic time.

  Assumes (i < weight tree).
 *)
let rec extract i ((w, t) : 'a weighted_tree) : 'a weighted * 'a weighted_tree =
  assert (i < w) ;
  match t with
  | Nothing ->
      assert false (* impossible if i >= 0 *)
  | Leaf v ->
      ((w, v), nothing)
  | Node (left, right) ->
      let w_left = weight left in
      let elem, result =
        if i < w_left then
          let elem, left = extract i left in
          (elem, node left right)
        else
          let elem, right = extract (i - w_left) right in
          (elem, node left right)
      in
      assert (weight result + weight elem = w) ;
      (elem, result)

(* [consume_list n li] builds a well-balanced tree from the n first elements of li;
   assumes (n <= List.li) *)
let rec consume_list (n : int) (li : 'a weighted list) : 'a weighted_tree * 'a weighted list =
  match (n, li) with
  | 0, _ ->
      (nothing, li)
  | 1, x :: xs ->
      (leaf x, xs)
  | 1, [] ->
      assert false (* we assume n <= List.length li *)
  | _, _ ->
      let left, li = consume_list (n / 2) li in
      let right, li = consume_list (n - (n / 2)) li in
      (node left right, li)

let of_list li =
  let tree, li = consume_list (List.length li) li in
  assert (li = []) ;
  tree

(* [random_weighted rnd li] is the result of repeatedly picking (and removing)
   elements of [li] chosen at random according to their weight.

   The output list may be smaller than the input list, as elements with
   weight 0 are not included -- they are never picked. *)
let random_weighted (rnd : int -> int) (li : 'a weighted list) : 'a list =
  let rec to_random_list t =
    match weight t with
    | 0 ->
        []
    | w ->
        let (_, v), t = extract (rnd w) t in
        v :: to_random_list t
  in
  to_random_list (of_list li)
```

I would propose to finish reviewing your version and merge it, as a good baseline for performance. Then we could (would you be interested in doing it? otherwise I would) try this implementation instead. It may be faster, but it may also be slower: if the lists of rules are very short, the additional cost of building the weighted tree may offset the complexity benefit.

ulugbekna commented:
>  I'm not completely convinced by the comment style

Agreed. I liked the style of comments in your implementation of the `gen_term_from_rules`. I will try to adhere to that style and follow your recommendations.

Regarding your implementation, I liked it a lot and will try to benchmark and see if we can get better running time.

gasche commented:
I mentioned that my code could be faster or slower, but one other possible outcome is that we don't see any difference -- in particular if there are differences in the random generator behavior, the difference would be hard to measure. In particular, it could be that one implementation is in fact faster than the other, but that the differences are drowned out by other sources of inefficiency somewhere else in the generator. (Profiling this code is not very easy because it's very monadic, with most of the time spent in chaining un-named function closures.)

In theory, all three functions (the old implementation, yours and mine) need the same number of random-generator calls, and it would be possible that the implementations make exactly the same choices given the same generator state (this might be the case already -- I would be surprised -- or it could require some tweaking). If that was the case, then it would be easy to measure performance differences, by just using a fixed seed. Do you know if your implementation makes the same random choices as the previous one? (From a fixed seed, does it generate the same programs?)

gasche commented:
Partly answering my own question: the implementation of `Gen.frequencyl` computes the sum of all weights just as you do, and selects the element from the list in a way that is very close to what you do, but the computation of the index is different:

- you use `Gen.int_bounds w_sum`, which returns an integer within `[0..w_sum]`
- `Gen.frequencly` uses `Random.State.int w_sum`, which makes different random draws (more on that below) and returns an integer within `[0 .. w_sum - 1]`, bound excluded.

I think that excluding the bound is the correct choice. If you have a list of two elements with weights 1, you want to draw in `[0; 1]` and select either of them, instead of drawing in `[0; 2]` and selecting the second for both `1` and `2`. So you should probably call `Gen.int_bound (w_sums - 1)` in any case.

The difference between the implementations `Random.State.int` and `Gen.int_bound` means that, even if the issue above is fixed, the two implementations will affect the random-generator state in different ways. I would recommend using `(fun st -> Random.State.int st w_sums)` here (and also if/when you try to use my implementation), if that can get us the exact same choice, as it would make comparing the three implementations much easier. (The code is more low-level and less pleasant, but we can always revert to `Gen.int_bound` later.)

gasche commented:
Thinking more about this: I think it would be nice to send a Pull Request to QCheck to change the implementation of `int_bound` so that, when the given bound is smaller than 2^30 (`bound <= 1 lsl 30 - 1`)¹, it uses `Random.int` exactly instead of its own logic. In our case, for example, this would guarantee that `frequencyl` and our own reimplementation of it, using the primitives provided by `QCheck`, have the exact same behavior.

¹: I think that one has to write `bound <= 1 lsl 30 - 1` instead of `bound < 1 lsl 30` because, on 32-bit architectures, the two expressions are not equivalent (`1 lsl 30 - 1` is `max_int`, and `1 lsl 30` is `min_int`, so it is negative).

gasche commented:
Of course the internet has a [much shorter solution](https://softwareengineering.stackexchange.com/a/344274/33960) to the problem of weighted random shuffle, whose OCaml translation is the following:

```ocaml
let random_weighted (li : (float * 'a) list) : 'a list =
  let sample (w, v) = (Random.float 1. ** (1. /. w), v) in
  let samples = List.map sample li in
  List.sort (fun (w1, _) (w2, _) -> compare w2 w1) samples
  |> List.map snd
```

This implementation is likely to be faster than all others, so we could just use it directly (adapted to take integer weights in our case).

(In fact we could suggest this function for inclusion in QCheck.)

jmid commented:
This is very nice @gasche !
1. QCheck is missing a float generator with output in [0;1] (or [0;1[) as far as I remember.
   Adding that, and then the above would be a nice contribution indeed.

2. I noticed that the above algorithm computes a permution/shuffle rather than "just a selection".
    Of course, if you can shuffle, then you can also select: shuffle first, then select the head.
    More seriously, the shuffle may be give us a hint to a better/alternative/faster generation strategy:
    rather than 
      - randomly select, try selected generator, if it fails remove generator and start over
         (which repeatedly generates) we could
      - randomly shuffle (using the above), the outcome determines the sequence we will try things in:
         if the first generator fails, try the next, etc.
        This way generation is contained to one point: the shuffle.

3. I still need to convince myself that the above gives the expected distribution for weights different
    from 1.0 (but I was probably just up too early today...) 😄

gasche commented:
Re. contributions to QCheck: yes, we highlighted some potential contributions above. I would like to give @ulugbekna the opportunity to contribute them, if he is interested.

Re. the probability stuff. Here is how I convinced myself. (I had re-learn some bits of probability theory to do it, so it may very well be wrong).

I prove that if we have (w1, e1) and (w2, e2) in the list, then the probability of seeing e1 *after* e2 in the resulting shuffle is (w2 / (w1+w2)), which sounds like exactly what you expect from a random weighted shuffle. (If w2 is much bigger than w1, this probability is large.)  
Let us call x1 the random[0;1] number drawn for (w1, e1), and x2 for (w2, e2). The probability of seing e1 after e2 is the probability that `x1 ^ (1/w1) < x2 ^ (1/w2)`. (The sorting function puts the largest values *first* in the result, which explains the direction of the inequality.)   
By putting both sides to the (positive) power w1, this is the probability that `x1 < x2 ^ (w1/w2)`.  
For a fixed `x2`, this probability is exactly `x2 ^ (w1/w2)`, as this is itself in [0;1].  
So the probability for any `x2` is the integral of all those probabilities `x2 ^ (w1/w2)`, for any x2 in [0;1].  
The integral of this formula is (1 / (w1/w2 + 1)), which is (1 / ((w1+w2) / w2)), which is `w2 / (w1+w2)`.

gasche commented on src/efftester.ml:
```diff
@@ -1089,39 +1089,47 @@ module Generators (Ctx : Context) = struct
+      | [] -> Gen.return None
+      | rls ->
+        let open Gen in
+        bound_int_gen w_sum >|= gen_freq rls >>= fun ((w, g), new_rls) ->
```
this should be `bound_int_gen (w_sum - 1)`, but otherwise this is good to go.

ulugbekna replied:
But we are using `Random.State.int` which already excludes the upper boundary. Am I missing something?

gasche replied:
(indeed; see comments in the discussion. Marking this as resolved.)

ulugbekna commented:
Thank you @gasche for the rigorous approach to the problem! I fixed up all problems that the current PR had. (The most difficult part was to come up with a comment that is concise and useful!) 

I will add to the to-do list:
- making a PR to the QCheck (exciting!)
- changing the shuffle implementation to O(nlogn) one

I chose to decrement `rand_k` in the first version of `gen_freq` because of the difference in boundary inclusion in `int_bound` and `Random.State.int`, as you indicated in one of your comments. It produced better distributions, but as we switched simply to `Random.State.int` generator, the problem disappeared. (Also, in my implementation `rand_k` could go negative, which I haven't thought of)

gasche commented:
Sorry, I had missed the use of `Random.State.int`. Could we maybe give it a name that is more different from the `Gen.int_bound` function which has a different specification (on the bound)? For example, `int_bound_excluded` could be clearer, or `int_below`.

gasche commented:
I liked the state reached by the PR so I merged the code.

Feel free to open a new PR for further changes to this function, in particular if you want to merge one of the n-log-n versions -- I would recommend going with the simplest version using floating-point operations.

We discussed several contributions to QCheck, and I would encourage you to consider doing any of them that you are interested in:

1. Changing the `int_bound` generator to behave just like `Random.State.int` on inputs below 2^30

2. Bringing a weighted-list-permutation generator to QCheck. Two interfaces would be possible, one is something like `shuffle_w_l : (int * 'a) list -> 'a list t` as a generalization of `shuffle_l` (then our use-case, where the things to select are themselves generators, requires a bind on the user side), the other would be something like `try_frequency : (int * 'a option t) list -> 'a option t`,, which randomly tries *partial* generators and returns a value from the first one that suceeds -- this requires shuffling the list internally.

3. Having a floating-point generator that returns a uniform value between `0.` and a bound argument. Personally I would propose both `float_bound_included` and `float_bound_excluded`, and also have the same with `int_bound_{included,excluded}`, but I don't know whether the QCheck author woud like that.

I realized while re-reading the QCheck interface for this discussion that QCheck already has a parametrized fixpoint combinator (previously I mentioned, in our float-gen discussion, that it was missing, but it is specialized to the `'a sized` case where the parameter is an integer. I will send a QCheck PR of my own to generalize this interface to arbitrary parameters.

ulugbekna commented:
I have started working on the three PRs to QCheck. I have some questions: 

1) I share your POV regarding the implementation of `int_bound`. However, QCheck has its own implementation of `int_bound`. Isn't it for a reason?

2) I am planning on going with `shuffle_w_l`.

3) I have started working with this first. I am exploring how I can get a bounded float gen.

gasche commented:
Re. (1): `Random.int` only handles bound below 2^30, and I think the QCheck authors wanted to ensure that they would support *any* input bound (in particular: from 2^30 to 2^62 on 64bits machines), so they decided to reimplement their own. I think that it is *possible* to have your cake and eat it too, by ensuring that you behave exactly as `Random.int` (by calling it directly) on small inputs (or, on 32-bit machines, on all inputs), and then do  your own thing for the rest. But you should feel free to ask the QCheck maintainers directly if you wonder about it, maybe they have a different answer
